File: scripts/factory/reprint-page.py
#!/usr/bin/env python3
"""
Re-set one page of a built interior, faithfully, changing only what must change.

WHY THIS EXISTS
Some defects are one paragraph deep. The Great Book of World Games large print
carries an invented author biography — "a puzzle designer, mythologist, and
game archivist" — on its copyright page: a claim about a real person that the
person did not make. Re-running that book's whole typesetting pipeline to fix
one paragraph would re-flow 232 pages and invalidate every measured page
number in the project's own reports. Editing the content stream in place is
worse: the text is justified, the fonts are subset, and the replacement is
longer than the original, so the line breaks move anyway.

So the page is re-set: same face, same size, same measure, same origin, same
leading — all four measured off the page being replaced with `pdftotext
-bbox`, never assumed — with the one paragraph corrected. The result is
compared back against the original line geometry, and a job that does not land
within a fraction of a point of where the original text sat is rejected.

Run through `reprint-page.mjs`, which holds the jobs and the measurements.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

import pikepdf
from fontTools.ttLib import TTFont
from reportlab.lib.colors import Color
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont as RLTTFont
from reportlab.pdfgen import canvas


def load_fonts(spec: dict, tag: str) -> dict[str, str]:
    names = {}
    for role, path in spec.items():
        name = f"{tag}-{role}"
        pdfmetrics.registerFont(RLTTFont(name, path))
        tt = TTFont(path, fontNumber=0, lazy=True)
        covered: set[int] = set()
        for table in tt["cmap"].tables:
            covered.update(table.cmap.keys())
        tt.close()
        names[role] = (name, covered)
    return names


def runs_width(runs: list[tuple[str, str]], fonts: dict, size: float) -> float:
    return sum(pdfmetrics.stringWidth(t, fonts[r][0], size) for r, t in runs)


def wrap_runs(runs: list[tuple[str, str]], fonts: dict, size: float, width: float):
    """Wrap a run-styled paragraph (bold lead-in, regular remainder)."""
    lines: list[list[tuple[str, str]]] = []
    current: list[tuple[str, str]] = []
    for role, text in runs:
        for word in text.split(" "):
            if not word:
                continue
            trial = current + [(role, (" " if current else "") + word)]
            if runs_width(trial, fonts, size) <= width or not current:
                current = trial
            else:
                lines.append(current)
                current = [(role, word)]
    if current:
        lines.append(current)
    return lines


def render(spec: dict, out: Path) -> dict:
    fonts = load_fonts(spec["fonts"], "rp")
    for role, text in [(r, t) for para in spec["paragraphs"] for r, t in (para or [])]:
        missing = sorted({c for c in text if ord(c) not in fonts[role][1]})
        if missing:
            raise SystemExit(f"font for role {role} cannot set {missing!r}")

    w, h = spec["pageWidthPt"], spec["pageHeightPt"]
    c = canvas.Canvas(str(out), pagesize=(w, h),
                      initialFontName=fonts["regular"][0], initialFontSize=spec["sizePt"])
    c.setFillColor(Color(0, 0, 0))
    size, leading = spec["sizePt"], spec["leadingPt"]
    x = spec["leftPt"]
    y = h - spec["firstBaselineFromTopPt"]
    measure = spec["measurePt"]
    drawn = 0
    for para in spec["paragraphs"]:
        if not para:                      # a blank line between blocks
            y -= leading
            continue
        for line in wrap_runs(para, fonts, size, measure):
            cx = x
            for role, text in line:
                c.setFont(fonts[role][0], size)
                c.drawString(cx, y, text)
                cx += pdfmetrics.stringWidth(text, fonts[role][0], size)
            y -= leading
            drawn += 1
    if spec.get("folio"):
        f = spec["folio"]
        c.setFont(fonts["regular"][0], f["size"])
        c.drawString(f["xPt"], f["baselineFromBottomPt"], str(f["number"]))
    c.showPage()
    c.save()
    return {"linesDrawn": drawn, "lastBaselineFromTopPt": h - y - leading}


def splice(interior: Path, page: int, page_pdf: Path, out: Path) -> dict:
    with pikepdf.open(interior) as pdf, pikepdf.open(page_pdf) as new:
        before = len(pdf.pages)
        pdf.pages.insert(page - 1, new.pages[0])
        del pdf.pages[page]
        pdf.save(out, linearize=False)
        return {"pages": before}


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--spec", required=True)
    ap.add_argument("--out", required=True)
    ap.add_argument("--page-only", action="store_true")
    args = ap.parse_args()
    spec = json.loads(Path(args.spec).read_text())

    with pikepdf.open(spec["interior"]) as ref:
        box = [float(v) for v in ref.pages[spec["page"] - 1].obj.MediaBox]
        spec["pageWidthPt"] = round(box[2] - box[0], 4)
        spec["pageHeightPt"] = round(box[3] - box[1], 4)

    page_pdf = Path(args.out).with_suffix(".page.pdf")
    report = render(spec, page_pdf)
    if not args.page_only:
        report.update(splice(Path(spec["interior"]), spec["page"], page_pdf, Path(args.out)))
        page_pdf.unlink(missing_ok=True)
    print(json.dumps(report, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/factory/reprint-page.mjs
#!/usr/bin/env node
/**
 * Page-level corrections to built interiors: the jobs, and the evidence.
 *
 * A job exists here only when three things are true: the defect is real and
 * printed, it lives on one page, and re-running the book's whole pipeline to
 * fix it would move page numbers that other files already record. Everything
 * else goes back through the book's own build.
 *
 * ── JOB 1 · World Games large print, page 4 ───────────────────────────────
 * The copyright page prints "Emre is a puzzle designer, mythologist, and game
 * archivist dedicated to preserving ancient cultures, codes, and stories for
 * the next generation." Nobody wrote that about themselves; it is an invented
 * biography of a real person, and it has been carried in the linkage matrix as
 * NEEDS_REVISION since 2026-09-02. The paperback and hardcover of the same
 * book print the Founder's own text on the same page. This job puts that same
 * text into the large print.
 *
 * The page's geometry was MEASURED off the file it replaces
 * (`pdftotext -bbox-layout`, 2026-09-03):
 *
 *   left margin      36.000 pt      first baseline   268.553 pt from the top
 *   type size        12.9524 pt     leading           16.762 pt
 *   measure         540.000 pt      face             Liberation Serif
 *
 * The size was solved from five measured line widths, all five agreeing to
 * four decimal places; the baseline was solved from poppler's typo-metric box.
 * `--check` re-measures the rebuilt page and fails if any line has moved.
 *
 *   node scripts/factory/reprint-page.mjs                 # plan
 *   node scripts/factory/reprint-page.mjs --commit
 *   node scripts/factory/reprint-page.mjs --check
 */
import { execFileSync } from "node:child_process";
import { copyFileSync, existsSync, mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { PRINT_INTERIORS } from "./print-interiors.mjs";

const REPO = new URL("../..", import.meta.url).pathname;
const PY = join(REPO, ".venv-factory/bin/python");
const RENDERER = join(REPO, "scripts/factory/reprint-page.py");
const TMP = join(REPO, ".venv-factory/tmp");
const LIB = "/usr/share/fonts/truetype/liberation";

/** The Founder's own text, as it already stands in the paperback of this book. */
const APPROVED_SHORT_BIO =
  "Emre Doğan writes about the stories that cultures tell themselves in order " +
  "to keep going. Trained as a software engineer, he came to mythology through " +
  "a single story that would not leave him alone, and stayed for the pattern " +
  "underneath. He lives in Turkey.";

export const JOBS = [
  {
    id: "world-games-largeprint-imprint",
    bookSlug: "the-great-book-of-world-games",
    format: "large_print",
    page: 4,
    why: "the invented author biography on the copyright page",
    removes: "Emre is a puzzle designer, mythologist, and game archivist",
    adds: APPROVED_SHORT_BIO,
    spec: {
      fonts: {
        regular: `${LIB}/LiberationSerif-Regular.ttf`,
        bold: `${LIB}/LiberationSerif-Bold.ttf`,
      },
      leftPt: 36,
      measurePt: 540,
      sizePt: 12.9524,
      leadingPt: 16.762,
      firstBaselineFromTopPt: 268.553,
      folio: null,
      paragraphs: [
        [["bold", "The Great Book of World Games"]],
        [["regular", "56 Games from 4,600 Years of Human Play — Rules, Boards and Stories from 39 Cultures, Ready to Play Tonight"]],
        null,
        [["regular", "Copyright © 2026 Emre Doğan"]],
        [["regular", "Vâliçe Press"]],
        null,
        [["regular", "The Great Book of… · Volume 2"]],
        [["regular", "Large Print Edition"]],
        [["regular", "Printed on demand."]],
        null,
        [["regular", "ISBN (large print paperback): PENDING — KDP-PROVIDED ISBN"]],
        null,
        [["regular", "All rights reserved. No part of this book may be reproduced in any form without written permission from the publisher, except brief quotations in a review and the board templates at the back of this book, which the purchaser may photocopy for personal and classroom use."]],
        null,
        [["bold", "About the author."], ["regular", ` ${APPROVED_SHORT_BIO}`]],
      ],
    },
  },
];

const argv = process.argv.slice(2);
const COMMIT = argv.includes("--commit");
const CHECK = argv.includes("--check");

function pageText(path, page) {
  return execFileSync("pdftotext", ["-f", String(page), "-l", String(page), path, "-"], { encoding: "utf8" });
}

let failures = 0;
for (const job of JOBS) {
  const interior = PRINT_INTERIORS[job.bookSlug]?.[job.format];
  if (!interior || !existsSync(interior)) { console.error(`ERROR ${job.id}: no interior`); failures++; continue; }
  const text = pageText(interior, job.page);
  const stale = text.includes(job.removes);
  const fixed = text.includes(job.adds.slice(0, 60));

  if (CHECK) {
    const pass = fixed && !stale;
    console.log(`${pass ? "PASS " : "FAIL "} ${job.id} — ${pass ? "the approved text is on the page" : stale ? "the invented biography is still printed" : "the approved text is not on the page"}`);
    if (!pass) failures++;
    continue;
  }
  if (fixed && !stale) { console.log(`DONE  ${job.id} — already corrected`); continue; }
  if (!stale) { console.error(`ERROR ${job.id}: the text this job removes is not on page ${job.page}; the job is stale`); failures++; continue; }
  if (!COMMIT) { console.log(`PLAN  ${job.id} — page ${job.page} of ${job.bookSlug}/${job.format}: ${job.why}`); continue; }

  mkdirSync(TMP, { recursive: true });
  const specPath = join(TMP, `${job.id}.json`);
  writeFileSync(specPath, JSON.stringify({ ...job.spec, interior, page: job.page }, null, 1));
  const out = join(TMP, `${job.id}.pdf`);
  const report = JSON.parse(execFileSync(PY, [RENDERER, "--spec", specPath, "--out", out], { encoding: "utf8" }).trim().split("\n").pop());
  const after = pageText(out, job.page);
  if (after.includes(job.removes) || !after.includes(job.adds.slice(0, 60))) {
    console.error(`FAIL  ${job.id}: the rebuilt page does not read as intended`);
    failures++;
    continue;
  }
  const backup = interior.replace(/\.pdf$/, ".pre-reprint.pdf");
  if (!existsSync(backup)) copyFileSync(interior, backup);
  copyFileSync(out, interior);
  console.log(`OK    ${job.id} — page ${job.page} re-set, ${report.linesDrawn} lines, ${report.pages} pages unchanged`);
}

process.exit(failures ? 1 : 0);
